<center style='font-size:28px;'><u><b>SQL Task</b></u></center>

The goal of this study is to generate a value proposition for a new product using a database of one of the services competing in the market, which contains data on books, publishers, authors, customer ratings and reviews of books.

# Database Connection Setup <a id='dcs'></a>

In [1]:
# pip install psycopg2

In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Looking at the Data <a id='latd'></a>

## Books <a id='bk'></a>

In [3]:
query = ''' SELECT *
            FROM books    
        '''

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

There are 1000 books in the database.

In [4]:
query = ''' SELECT book_id, author_id, title, num_pages, publication_date, publisher_id, COUNT(*)
            FROM books
            GROUP BY book_id, author_id, title, num_pages, publication_date, publisher_id
            HAVING COUNT(*) > 1    
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id, count]
Index: []

In [22]:
query = ''' SELECT *
            FROM books
            WHERE book_id IS NULL OR author_id IS NULL OR title IS NULL OR
            num_pages IS NULL OR publication_date IS NULL OR publisher_id IS NULL
            '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

No duplicates or missing values

## Authors <a id='ath'></a>

In [5]:
query = ''' SELECT *
            FROM authors    
        '''

pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [23]:
query = ''' SELECT author_id, author, COUNT(*)
            FROM authors
            GROUP BY author_id, author
            HAVING COUNT(*) > 1    
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [author_id, author, count]
Index: []

In [24]:
query = ''' SELECT * 
            FROM authors
            WHERE author_id IS NULL OR author IS NULL
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [author_id, author]
Index: []

636 authors. No duplicates or missing values

## Publishers <a id='pbl'></a>

In [6]:
query = ''' SELECT *
            FROM publishers    
        '''

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

340 publishers

In [25]:
query = ''' SELECT publisher_id, publisher, COUNT(*)
            FROM publishers
            GROUP BY publisher_id, publisher
            HAVING COUNT(*) > 1    
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [publisher_id, publisher, count]
Index: []

In [26]:
query = ''' SELECT *
            FROM publishers
            WHERE publisher_id IS NULL OR publisher IS NULL
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

No duplicates or missing values

## Ratings <a id='rtng'></a>

In [7]:
query = ''' SELECT *
            FROM ratings    
        '''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

6456 ratings

In [30]:
query = ''' SELECT rating_id, book_id, username, rating, COUNT(*)
            FROM ratings
            GROUP BY rating_id, book_id, username, rating
            HAVING COUNT(*) > 1    
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [rating_id, book_id, username, rating, count]
Index: []

In [31]:
query = ''' SELECT *
            FROM ratings
            WHERE rating_id IS NULL OR book_id IS NULL OR username IS NULL OR rating IS NULL
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

No duplicates or missing values

In [8]:
query = ''' SELECT COUNT(DISTINCT username)
            FROM ratings
        '''

pd.io.sql.read_sql(query, con = engine)

count
0    160

Performed by 160 users.  Around 40 ratings for a rating-user

In [9]:
query = ''' SELECT AVG(rating)
            FROM ratings
        '''

pd.io.sql.read_sql(query, con = engine)

avg
0  3.928284

The average rating is 4.

## Reviews <a id='rvw'></a>

In [10]:
query = ''' SELECT *
            FROM reviews    
        '''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

And 2793 reviews

In [32]:
query = ''' SELECT review_id, book_id, username, text, COUNT(*)
            FROM reviews
            GROUP BY review_id, book_id, username, text
            HAVING COUNT(*) > 1    
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text, count]
Index: []

In [33]:
query = ''' SELECT *
            FROM reviews
            WHERE review_id IS NULL OR book_id IS NULL OR username IS NULL OR text IS NULL
        '''

pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

In [ ]:
No duplicates or missing values.

In [11]:
query = ''' SELECT COUNT(DISTINCT username)
            FROM reviews
        '''

pd.io.sql.read_sql(query, con = engine)

count
0    160

Again, performed by 160 users. Around 17 reviews for a reviewing-user

## Its The Same 160 Users <a id='itsu'></a>

In [12]:
query = ''' WITH a AS (
            SELECT DISTINCT username
            FROM reviews),
            b AS (SELECT DISTINCT username
            FROM ratings)
            
            SELECT a.username, b.username
            FROM a
            JOIN b
            ON a.username = b.username
        '''

pd.io.sql.read_sql(query, con = engine)

username        username
0         zjohnston       zjohnston
1           jnelson         jnelson
2      lewisdesiree    lewisdesiree
3     patrickhudson   patrickhudson
4             ulowe           ulowe
..              ...             ...
155  shermannatalie  shermannatalie
156         dmiller         dmiller
157  charlesraymond  charlesraymond
158     isaiahreyes     isaiahreyes
159       masonsara       masonsara

[160 rows x 2 columns]

So, those 2 sets of users are the same

# Preliminary Notes <a id='prlmnrn'></a>

So for this data -
- No duplicates or missing values
- Data contains 1000 books
- Average author wrote 1.5 books
- Average publisher published 3 books
- Average book has 6 ratings and 3 reviews
- 160 users performed reviews and ratings (average of 17, 40)

# Task 1 <a id='1'></a>

In [13]:
query = ''' SELECT COUNT(book_id) AS n_books
            FROM books
            WHERE publication_date > '2000-01-01'
        '''

pd.io.sql.read_sql(query, con = engine)

n_books
0      819

819 books published in the 21st century, 82% of the data's books

# Task 2 <a id='2'></a>

In [14]:
query = ''' WITH a AS (SELECT reviews.book_id,
            COUNT(review_id),
            AVG(rating)
            FROM reviews
            JOIN ratings ON reviews.book_id = ratings.book_id
            GROUP BY reviews.book_id)
            
            SELECT books.title,
            a.count AS n_reviews,
            a.avg AS avg_rating
            FROM a
            LEFT JOIN books
            ON a.book_id = books.book_id
            ORDER BY n_reviews desc, avg_rating DESC
        '''

pd.io.sql.read_sql(query, con = engine)

title  n_reviews  avg_rating
0                              Twilight (Twilight  #1)       1120    3.662500
1                  The Hobbit  or There and Back Again        528    4.125000
2                               The Catcher in the Rye        516    3.825581
3    Harry Potter and the Prisoner of Azkaban (Harr...        492    4.414634
4    Harry Potter and the Chamber of Secrets (Harry...        480    4.287500
..                                                 ...        ...         ...
989                                Moo  Baa  La La La!          2    3.000000
990                     Death: The High Cost of Living          2    3.000000
991                                  History of Beauty          2    2.500000
992                 1 000 Places to See Before You Die          2    2.500000
993               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1          2    2.500000

[994 rows x 3 columns]

The most reviewed book was reviewed 66 times the average! 6 books are missing tough:

In [15]:
query = ''' SELECT DISTINCT book_id
            FROM ratings
        '''

pd.io.sql.read_sql(query, con = engine)

book_id
0        652
1        273
2         51
3        951
4        839
..       ...
995       64
996       55
997      148
998      790
999      828

[1000 rows x 1 columns]

They all have rating

In [16]:
query = ''' SELECT DISTINCT book_id
            FROM reviews
        '''

pd.io.sql.read_sql(query, con = engine)

book_id
0        652
1        273
2         51
3        951
4        839
..       ...
989       64
990       55
991      148
992      790
993      828

[994 rows x 1 columns]

But all 6 don't have a review

In [17]:
query = '''
        WITH a AS (
        SELECT DISTINCT reviews.book_id
        FROM reviews
        JOIN books
        ON reviews.book_id = books.book_id),
        
        b AS (
        SELECT books.book_id,
               books.author_id,
               books.title,
               books.num_pages,
               books.publication_date,
               books.publisher_id
        FROM a
        JOIN books
        ON a.book_id = books.book_id),
        
        c AS (
        SELECT book_id
        FROM b),
        
        d AS (SELECT *
              FROM books
              WHERE book_id NOT IN(SELECT book_id FROM c)),
        
        e AS (SELECT title,
        AVG(rating) as rating
        FROM d
        JOIN ratings ON d.book_id = ratings.book_id
        GROUP BY title)
        

        SELECT *
        FROM e
        ORDER BY rating DESC
        '''
        
pd.io.sql.read_sql(query, con = engine)

title    rating
0   The Cat in the Hat and Other Dr. Seuss Favorites  5.000000
1                               Leonardo's Notebooks  4.000000
2                          Essential Tales and Poems  4.000000
3  Disney's Beauty and the Beast (A Little Golden...  4.000000
4    Anne Rice's The Vampire Lestat: A Graphic Novel  3.666667
5                            The Natural Way to Draw  3.000000

Those books weren't reviewed (`n_reviews` = 0)

# Task 3 <a id='3'></a>

In [18]:
query = ''' SELECT count(publishers.publisher),
            publishers.publisher
            FROM books join
            publishers on books.publisher_id = publishers.publisher_id
            where books.num_pages > 50
            group by publishers.publisher
            order by count desc
            limit 1
        '''

pd.io.sql.read_sql(query, con = engine)

count      publisher
0     42  Penguin Books

Penguin Books is the winner, with 42 published books (taking books with more than 50 pages). Average publisher published 3 books.

# Task 4 <a id='4'></a>

In [19]:
query = ''' with a as (SELECT books.book_id,
            count(rating_id) as ratings_per_book,
            avg(rating)
            FROM books join
            authors on books.author_id = authors.author_id
            join ratings on books.book_id = ratings.book_id
            group by books.book_id),
            
            s as (select * from a
            where a.ratings_per_book > 49
            order by a.avg desc
            limit 1)
            
            select author,
            title,
            avg as avg_rating
            from s
            left join books on s.book_id = books.book_id
            left join authors on books.author_id = authors.author_id
        '''

pd.io.sql.read_sql(query, con = engine)

author  \
0  J.K. Rowling/Mary GrandPré   

                                               title  avg_rating  
0  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634

Among books with at least 50 ratings, J.K. Rowling is the author with the highest average book rating (4.4), for "Harry Potter and the Prisoner of Azkaban". Mary GrandPré is the illustrator. Average rating in the data is 4.

# Task 5 <a id='5'></a>

In [20]:
query = ''' with a as (select username,
        count(distinct book_id) 
        from ratings
        group by username
        order by count desc),
        
        b as (select *
        from a
        where count > 50),
        
        c as (select b.username,
        count(review_id)
        from b
        join reviews
        on b.username = reviews.username
        group by b.username)
        
        select avg(count)
        from c
        '''

pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

The average user, among those who rated more than 50 books, submitted 24 reviews. The average reviews per user is 17.